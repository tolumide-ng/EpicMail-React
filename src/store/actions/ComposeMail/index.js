import * as Toastr from 'toastr';
import axios from 'axios';

import config from '../../../config';
import {
  COMPOSE_PENDING,
  COMPOSE_SUCCESSFUL,
  COMPOSE_FAILURE
} from '../../actiontypes/messages';

export const composePending = () => ({
  type: COMPOSE_PENDING,
  payload: {
    composeStatus: 'pending',
    composeError: null,
    composeSuccess: null
  }
});

export const composeFailure = composeError => ({
  type: COMPOSE_FAILURE,
  payload: {
    composeStatus: 'fail',
    composeError,
    composeSuccess: null
  }
});

export const composeSuccessful = composeSuccess => ({
  type: COMPOSE_SUCCESSFUL,
  payload: {
    composeStatus: 'success',
    composeError: null,
    composeSuccess
  }
});

export const composeMailAction = ({
  recipient: receiverEmail,
  subject,
  message
}) => async dispatch => {
  dispatch(composePending());

  try {
    let user = localStorage.getItem('user');
    user = JSON.parse(user);
    const { token } = user;

    const response = await axios({
      method: 'POST',
      data: {
        receiverEmail: receiverEmail[0],
        subject: subject[0] || '',
        message: message[0] || ''
      },
      headers: {
        authorization: `bearer ${token}`
      },
      url: `${config.apiUrl}messages/`
    });

    const { message: data } = response.data;
    Toastr.success(data);

    dispatch(composeSuccessful(data));

    // const response = await fetch(`${config.apiUrl}messages/`, {
    //   method: 'post',
    //   headers: {
    //     'Content-type': 'application/x-www-form-urlencoded; charset=UTF-8',
    //     authorization: `bearer ${token}`
    //   },
    //   body: { receiverEmail, subject, message },
    //   mode: 'cors'
    // });
  } catch ({ response }) {
    const errorMessage = response.data.error || response;
    Toastr.error(errorMessage);
    dispatch(composeFailure(errorMessage));
  }
};
