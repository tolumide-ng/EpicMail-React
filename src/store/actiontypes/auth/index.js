export const SIGNUP_PENDING = 'SIGNUP_PENDING';

export const SIGNUP_FAILURE = 'SIGNUP_FAILURE';

export const SIGNUP_SUCCESS = 'SIGNUP_SUCCESS';

export const SET_USER = 'SET_USER';

export const LOGIN_PENDING = 'LOGIN_PENDING';

export const LOGIN_FAILURE = 'LOGIN_FAILURE';

export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';

export const REQUEST_RESET_PENDING = 'REQUEST_RESET_PENDING';

export const REQUEST_RESET_FAILURE = 'REQUEST_RESET_FAILURE';

export const REQUEST_RESET_SUCCESS = 'REQUEST_RESET_SUCCESS';

export const SIGNOUT_PENDING = 'SIGNUOUT_PENDING';

export const SIGNUOUT_SUCCESS = 'SIGNUOUT_SUCCESS';

export const SIGNOUT_FAILURE = 'SIGNOUT_FAILURE';
