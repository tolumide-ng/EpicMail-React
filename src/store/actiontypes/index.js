export const SIGNUP_PENDING = 'SIGNUP_PENDING';

export const SIGNUP_FAILURE = 'SIGNUP_FAILURE';

export const SIGNUP_SUCCESS = 'SIGNUP_SUCCESS';

export const LOGIN_PENDING = 'LOGIN_PENDING';

export const LOGIN_FAILURE = 'LOGIN_FAILURE';

export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';

export const REQUEST_RESET_PENDING = 'REQUEST_RESET_PENDING';

export const REQUEST_RESET_FAILURE = 'REQUEST_RESET_FAILURE';

export const REQUEST_RESET_SUCCESS = 'REQUEST_RESET_SUCCESS';
