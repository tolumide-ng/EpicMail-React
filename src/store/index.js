import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import rootReducer from './reducers';


const middlewares = applyMiddleware(thunk);

const store = createStore(
  rootReducer,
  composeWithDevTools(middlewares)
  );


export default store;