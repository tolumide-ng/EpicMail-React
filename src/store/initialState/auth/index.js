export const auth = {
  status: 'rest',
  error: null,
  isStarting: true,
  user: {}
};

export default auth;
