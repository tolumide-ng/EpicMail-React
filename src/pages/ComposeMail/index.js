import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as Toastr from 'toastr';

import { composeMailAction } from '../../store/actions/composeMail';
import { composeDraftAction } from '../../store/actions/composeDraft';
const regEmail = /^([a-zA-Z0-9_\-\.]+)@([a-zA-Z0-9_\-\.]+)\.([a-zA-Z]{2,5})$/;
import { checkLocalStorage } from '../../utils';

export class ComposeMail extends Component {
  state = {
    recipient: '',
    subject: '',
    message: '',
    errors: {
      recipient: '',
      subject: '',
      message: ''
    }
  };

  errorMessage = [];

  componentDidMount() {
    checkLocalStorage({ history: this.props.history });
  }

  validateForm = ({ errors, recipient }) => {
    let valid = true;
    Object.values(errors).forEach(
      val => val.length > 0 && this.errorMessage.push(val) && (valid = false)
    );

    recipient.length < 1 &&
      this.errorMessage.push('Recipient email cannot be empty') &&
      (valid = false);
    return valid;
  };

  saveDraft = e => {
    e.preventDefault();
    this.props.composeDraft({
      recipient: this.state.recipient,
      message: this.state.message,
      subject: this.state.message
    });
  };

  handleSubmit = e => {
    e.preventDefault();
    if (!this.validateForm(this.state)) {
      Toastr.error(this.errorMessage[this.errorMessage.length - 1]);
      return;
    }
    this.props.composeMail({
      recipient: this.state.recipient,
      message: this.state.message,
      subject: this.state.subject
    });
  };

  handleChange = e => {
    e.preventDefault();
    const { name, value } = e.target;
    let formErrors = this.state.errors;
    switch (name) {
      case 'recipient':
        formErrors.recipient =
          value.endsWith('@epicmail.com') && value.length > 0
            ? ''
            : 'Recipient email must end with @epicmail.com';
        break;
      default:
        break;
    }
    this.setState({ formErrors, [name]: [value] });
  };

  render() {
    const { recipient: recipientError } = this.state.errors;
    return (
      <>
        <div className="flex form_body m-10 mx-auto flex-col p-10 w-10/12 rounded">
          <div className="w-full bg-blue-700 text-base tracking-wide p-2 text-white">
            Compose Message
          </div>
          <div className="mt-4">
            <form onSubmit={this.handleSubmit}>
              <strong>From: </strong>
              <span>{this.props.user.email}</span>
              <div>
                <label htmlFor="recipient">
                  <input
                    placeholder="recipient@epicmail.com"
                    name="recipient"
                    type="email"
                    onChange={this.handleChange}
                    value={this.state.recipient}
                    className="mt-4 w-full h-10 p-2 pl-4 border border-gray-400 text-sm outline-none rounded-lg"
                  />
                  <div className="text-xs text-red-500 italic font-bold">
                    {recipientError.length > 1 && recipientError}
                  </div>
                </label>
              </div>
              <div>
                <label htmlFor="subject">
                  <input
                    placeholder="Subject"
                    name="subject"
                    type="text"
                    value={this.state.subject}
                    onChange={this.handleChange}
                    className="mt-4 w-full h-10 p-2 pl-4 border border-gray-400 text-sm outline-none rounded-lg"
                  />
                </label>
              </div>
              <div>
                <label htmlFor="message">
                  <textarea
                    placeholder="Your email here"
                    name="message"
                    type="text"
                    value={this.state.message}
                    onChange={this.handleChange}
                    rows="10"
                    cols="50"
                    className="mt-4 w-full p-2 pl-4 border border-gray-400 text-sm outline-none rounded-lg"
                  />
                </label>
              </div>
              <div className="flex justify-between mt-4">
                <button
                  onClick={this.saveDraft}
                  type="button"
                  name="draft"
                  className="bg-blue-500 text-white p-1 focus:bg-blue-700"
                >
                  Save as draft
                </button>
                <input
                  name="send"
                  type="submit"
                  value="Send"
                  className="bg-blue-500 text-white p-1 focus:bg-blue-700"
                />
              </div>
            </form>
          </div>
        </div>
      </>
    );
  }
}

const mapStateToProps = state => ({
  user: state.authReducer.user
});

const mapDispatchToProps = dispatch => ({
  composeMail: wholeMessage => dispatch(composeMailAction(wholeMessage)),
  composeDraft: wholeMessage => dispatch(composeDraftAction(wholeMessage))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ComposeMail);

// export default ComposeMail;
