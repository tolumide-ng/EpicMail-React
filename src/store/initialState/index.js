import { auth } from './auth';
import { composeMail } from './composeMail';
import { composeDraft } from './composeDraft';
import { sentMessages } from './sentMessages';
import { requestReset } from './requestReset';
import { deleteMessage } from './deleteMessage';
import { specificSentMessage } from './specificSentMessage';
import { retractMessage } from './retractMessage';

const initialState = {
  auth,
  requestReset,
  composeMail,
  composeDraft,
  sentMessages,
  deleteMessage,
  specificSentMessage,
  retractMessage
};

export default initialState;
