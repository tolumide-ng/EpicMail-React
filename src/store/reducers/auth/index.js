import {
  LOGIN_PENDING,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  SIGNUP_PENDING,
  SIGNUP_SUCCESS,
  SIGNUP_FAILURE
} from '../../actiontypes/auth';

import initialState from '../../initialState';

const authReducer = (state = initialState.auth, { type, payload }) => {
  switch (type) {
    case LOGIN_PENDING:
      return {
        ...state,
        ...payload
      };
    case LOGIN_SUCCESS:
      return {
        ...state,
        ...payload
      };
    case LOGIN_FAILURE:
      return {
        ...state,
        ...payload
      };
    case SIGNUP_PENDING:
      return {
        ...state,
        ...payload
      };
    case SIGNUP_SUCCESS:
      return {
        ...state,
        ...payload
      };
    case SIGNUP_FAILURE:
      return {
        ...state,
        ...payload
      };
    case 'SET_USER':
      return {
        ...state,
        ...payload
      };
    default:
      return state;
  }
};

export default authReducer;
