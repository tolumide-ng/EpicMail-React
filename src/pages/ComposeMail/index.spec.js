import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, fireEvent, cleanup, wait } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ComposeMail, { mapStateToProps, mapDispatchToProps } from './index';
import store from '../../store';
import { auth as initialState } from '../../store/initialState/auth';

describe.only('Compose Mail Form', () => {
  afterEach(cleanup);
  beforeAll(() => {
    // ComposeMail.prototype.componentDidMount = () => {};
    const spy = jest.mock(ComposeMail.prototype, 'componentDidMount');
  });
  const checkLocalStorage = jest.fn();
  test('renders with default props', () => {
    const { getByPlaceholderText, asFragment } = render(
      <MemoryRouter>
        <ComposeMail store={store} checkLocalStorage={checkLocalStorage} />
      </MemoryRouter>
    );

    const input = getByPlaceholderText('recipient@epicmail.com');
    expect(input).toHaveAttribute('type', 'email');

    const subjectInput = getByPlaceholderText('subject');
    expect(subjectInput).toHaveAttribute('type', 'text');

    const messageInput = getByPlaceholderText('Your email here');
    expect(messageInput).toHaveAttribute('type', 'text');
  });

  test('fires validation error when user input is not valid', async () => {
    const { getByPlaceholderText, asFragment, findByTestId } = render(
      <MemoryRouter>
        <ComposeMail store={store} checkLocalStorage={checkLocalStorage} />
      </MemoryRouter>
    );
    const emailInput = getByPlaceholderText('recipient@epicmail.com');

    fireEvent.change(emailInput, { target: { value: 'username' } });
    fireEvent.blur(emailInput);

    const validationErrors = await findByTestId('recipientError');
    expect(validationErrors.innerHTML).toBe(
      'Recipient email must end with @epicmail.com'
    );
  });

  test('does not fire validation error if there is no validation for the input field', async () => {
    const { asFragment, getByPlaceholderText, findByTestId } = render(
      <MemoryRouter>
        <ComposeMail store={store} />
      </MemoryRouter>
    );
    const input = getByPlaceholderText('subject');
    expect(asFragment).toMatchSnapshot();
    fireEvent.blur(input);
    // const validationErrors = await findByTestId('passwordError');
  });
});

describe.only('Dispatch actions on validation success', () => {
  afterEach(cleanup);

  test('Should submit form if register button is clicked', () => {
    const composeMail = jest.fn();
    const { getByPlaceholderText, getByText } = render(
      <MemoryRouter>
        <ComposeMail store={store} composeMail={composeMail} />
      </MemoryRouter>
    );
    const emailInput = getByPlaceholderText('recipient@epicmail.com');

    fireEvent.change(emailInput, {
      target: { persist: () => {}, value: 'theemail@email.com' }
    });
    fireEvent.blur(emailInput);
    fireEvent.click(getByText('Send'));
    // expect(composeMail).toHaveBeenCalled();
  });
});
