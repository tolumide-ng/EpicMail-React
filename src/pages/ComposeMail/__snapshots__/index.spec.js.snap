// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compose Mail Form does not fire validation error if there is no validation for the input field 1`] = `[Function]`;
